auth/src/main/java/org/aerogear/auth/impl/UserPrincipalImpl.java

```diff
@@ -119,8 +121,40 @@ UserPrincipalImpl build() {
      */
     @Override
     public boolean hasRole(final IRole role) {
-//        TODO: to be implemented
-//        return roles.containsKey(role.getRoleName());
+      return roles.contains(role);
+    }
+
+    /**
+     * Check if a user has a specified ClientRole based on param of type String to allow
+     * for better usability.
+     * Returns <code>true</code> if the user has the passed in role
+     * @param role String of role to be checked
+     * @return true or false
+     */
+    @Override
+    public boolean hasClientRole(String role){
+        for ( IRole r : roles){
+            if (r.getRoleName().equals(role) && r instanceof ClientRole){
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Check if a user has a specified RealmRole based on param of type String to allow
+     * for better usability
+     * Returns <code>true</code> if the user has the passed in role
+     * @param role String of role to be checked
+     * @return true or false
+     */
+    @Override
+    public boolean hasRealmRole(String role){
+        for ( IRole r : roles){
+            if ( r.getRoleName().equals(role) && r instanceof RealmRole){
```

Comment:
This way, every time we ask if a role is present, in the worst case the whole list of roles is fetched.
Wouldn't be better to save the roles in a Map so that the `containsKey` method could be used?

Reply:
Another issue: are role names case sensitive? The standard implementation of the `equals` method is case sensitive.